File: app/src/main/java/com/budgetbuddy/util/DateUtils.kt
package com.budgetbuddy.util

import java.text.SimpleDateFormat
import java.util.*

object DateUtils {

    fun startOfMonth(calendar: Calendar = Calendar.getInstance()): Long {
        return calendar.clone<Calendar>().apply {
            set(Calendar.DAY_OF_MONTH, 1)
            set(Calendar.HOUR_OF_DAY, 0)
            set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }.timeInMillis
    }

    fun endOfMonth(calendar: Calendar = Calendar.getInstance()): Long {
        return calendar.clone<Calendar>().apply {
            set(Calendar.DAY_OF_MONTH, getActualMaximum(Calendar.DAY_OF_MONTH))
            set(Calendar.HOUR_OF_DAY, 23)
            set(Calendar.MINUTE, 59)
            set(Calendar.SECOND, 59)
            set(Calendar.MILLISECOND, 999)
        }.timeInMillis
    }

    fun startOfDay(dateMs: Long = System.currentTimeMillis()): Long {
        return Calendar.getInstance().apply {
            timeInMillis = dateMs
            set(Calendar.HOUR_OF_DAY, 0)
            set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }.timeInMillis
    }

    fun endOfDay(dateMs: Long = System.currentTimeMillis()): Long {
        return Calendar.getInstance().apply {
            timeInMillis = dateMs
            set(Calendar.HOUR_OF_DAY, 23)
            set(Calendar.MINUTE, 59)
            set(Calendar.SECOND, 59)
            set(Calendar.MILLISECOND, 999)
        }.timeInMillis
    }

    fun startOfWeek(): Long {
        return Calendar.getInstance().apply {
            set(Calendar.DAY_OF_WEEK, firstDayOfWeek)
            set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0); set(Calendar.MILLISECOND, 0)
        }.timeInMillis
    }

    fun formatDate(dateMs: Long, pattern: String = "dd MMM yyyy"): String =
        SimpleDateFormat(pattern, Locale.getDefault()).format(Date(dateMs))

    fun formatAmount(amount: Double, currencySymbol: String = "R"): String =
        "$currencySymbol ${String.format("%.2f", amount)}"

    fun currentMonth(): Int = Calendar.getInstance().get(Calendar.MONTH) + 1
    fun currentYear(): Int  = Calendar.getInstance().get(Calendar.YEAR)
}

File: ios/BudgetBuddy/Utilities/DateUtils.swift
import Foundation

enum DateUtils {
    static func startOfMonth(_ date: Date = .now) -> Date {
        let cal   = Calendar.current
        let comps = cal.dateComponents([.year, .month], from: date)
        return cal.date(from: comps) ?? date
    }

    static func endOfMonth(_ date: Date = .now) -> Date {
        let cal   = Calendar.current
        var comps = cal.dateComponents([.year, .month], from: date)
        comps.month! += 1
        let nextMonth = cal.date(from: comps) ?? date
        return cal.date(byAdding: .second, value: -1, to: nextMonth) ?? date
    }

    static func startOfDay(_ date: Date = .now) -> Date {
        Calendar.current.startOfDay(for: date)
    }

    static func endOfDay(_ date: Date = .now) -> Date {
        Calendar.current.date(byAdding: .second, value: 86399, to: startOfDay(date)) ?? date
    }

    static var monthLabel: String {
        let fmt = DateFormatter()
        fmt.dateFormat = "MMMM yyyy"
        return fmt.string(from: .now)
    }
}
